### DISCUSSION TITLE
PageSpeed Performance

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3683] [4. General]
Summaried by Erwin Hofman [3 reasons why Alpine.js won't beat jQuery, Core Web Vitals included](https://www.erwinhofman.com/blog/3-reasons-why-alpinejs-wont-beat-jquery-core-web-vitals-included), I'm afraid to confirm that the PageSpeed score is rather low using AlpineJS (with only 1 additional plugin for i18n).

Are there some efforts in optimizing for speed/performance (and not only in size)?

**Comment by ekwoka** [comment]
It's already off to a bad start.

jQuery isn't beaten by Alpine. It's beaten by just normal JS.

JS already supersedes jQuery.

And there isn't any reason that AlpineJS would impact those performance metrics shown there.

It has performance issues for sure, mainly in the realm of large loops.

Other things are quite fast, though there are areas to improve it, but most of the cost just comes from the fact that you are writing expressions as just string values, not as actual JS code. It takes a bit of work to get around that.

**Comment by drewbitt** [comment]
It is well-known that Alpine has tons of rooms to grow regarding performance, as it has remained one of the [least performant JS frameworks for DOM manipulation](https://krausest.github.io/js-framework-benchmark/2023/table_chrome_115.0.5790.98.html). I have always been hoping this would improve before it goes into LiveWire 3, but it's looking more like after that.

**Reply by ekwoka** [reply]
There's a few fundamental aspects of the design of the language that can't really be gotten around, since it's JS expressions in string attributes. So extra work needs to be done just to handle that flow safely.

And then it is also a place where it's very easy to potentially create issues where it does way more work than needed, generally from fundamentally bad design or misplaced usage.

It think it's extremely good at the things it's really built to do. But there also aren't really any competitors in the specific niche Alpine operates in.

**Reply by sausawmg** [reply]
Does this mean that using Alpine.data is faster than using JS expressions in string attributes?

**Reply by ekwoka** [reply]
Yes, and even more specifically using x-bind to bulk bind Alpine behaviors from singular expressions.

But expressions on elements is the main reason for Alpine. That abstraction just has costs.

One way to improve it is to avoid the use of one off expressions when a common expression is possible.

Like if you are server rendering a list, and you want it to have reference to it's index or values in the expressions

```html
<div :class="selected = {{ item.title }} && 'bg-yellow-300">
```

This would generate a unique expression on every element, meaning many more functions.

Not an issue with small lists, but if something has 500 things, it can add up.

Changing it to be like

```html
<div :class="selected = $el.dataset.title && 'bg-yellow-300'" data-title="{{ item.title }}">
```

Would now make the function shared between all the elements.

But that does have some issues with the current implementations of element-bound magics, but a smaller degree I feel. Especially if you are already using magics anyway.

**Reply by simonferndriger** [reply]
Thanks for the hint.

Why is item.title in double curly brackets and what does it mean or how do you do server renerding this list with JS?

**Reply by ekwoka** [reply]
Thats jsut for common templating systems, like jinja, and django, and liquid, and mustache and such. Basically server rendered markup.

**Comment by simonferndriger** [comment]
Well, just because it's written in AlpineJS doesn't necessarily mean that
it needs to be delivered in AlpineJS. A (deluxe) solution to that problem
could be (for example) a plugin for the vite building process which
converts the AlpineJS code into native JS code for production.

Am Sa., 5. Aug. 2023 um 07:56 Uhr schrieb Eric Kwoka <
***@***.***>:

**Reply by ekwoka** [reply]
AlpineJS is already native JS. But yes, a potential way to increase performance in cases where some kinds of view renders are used is to compile the code in a manner that reduces the costs of handling function expressions on html elements.

Considering the main purpose of alpine is to use with LaravelLivewire, or others like Django, I don't think there is a decent way to meaningfully handle all of these.

**Comment by james0r** [comment]
This is interesting stuff and led me down a fun rabbit hole.

I've used AlpineJS for countless projects now and have not noticed any significant performance issues, but no doubt I'm not using it for any heavy lifting.

It sounds like performance issues arise in some known scenarios, so until things improve in those areas, it might be helpful to add some page/section to the docs that warns about or discourages certain known problem patterns.

Something like [https://mithril.js.org/lifecycle-methods.html#avoid-anti-patterns](https://mithril.js.org/lifecycle-methods.html#avoid-anti-patterns) perhaps.

**Reply by simonferndriger** [reply]
Yes, that would actually be helpful indeed

**Reply by ekwoka** [reply]
Yes, I've been thinking about PRing some docs updates from all my experience working on Alpine based sites to handle these.
